### PR TITLE
[#672] Fix handling X-Amz-Copy-Source header

### LINKED ISSUE #672: upload-part-copy created new part even we put wrong value to --copy-source parameter
**Steps to reproduce:**
1. Create source object
```
C:\TEMP>aws --no-verify-ssl s3api put-object --bucket b-test-lock-10 --key 'multipart-obj71' --body c:\TEMP\AWSCLIV2-download-p1.msi --endpoint-url http://172.26.163.38:8084
{
    "ETag": "54631d1173147c13e3c3c970ab19a449de07714ae6d1ec6f38a85bebdbf4f487"
}
```
2. Create multipartupload
```
C:\TEMP>aws s3api create-multipart-upload --bucket b-test-lock-10  --key 'multipart-obj70'  --endpoint-url http://172.26.163.38:8084
{
    "Bucket": "b-test-lock-10",
    "Key": "'multipart-obj70'",
    "UploadId": "aeb9bc9f-3a30-439d-af9a-894e150a1673"
}
```
3. Try to load source object as part of multipartupload with wrong  value of  --copy-source parameter. Parameter expected value "bucket/object"
```
C:\TEMP>aws s3api upload-part-copy --bucket b-test-lock-10  --key 'multipart-obj70'  --endpoint-url http://172.26.163.38:8084 --upload-id aeb9bc9f-3a30-439d-af9a-894e150a1673 --part-number 1 --copy-source multipart-obj71
An error occurred (Unknown) when calling the UploadPartCopy operation (reached max retries: 2): Unknown
```
4. As result part of object was created with size 0 bytes. But we expect that part shouldn't be created because we put the wrong value of  --copy-source parameter.
```
C:\TEMP>aws s3api list-parts --bucket b-test-lock-10  --key 'multipart-obj70'  --endpoint-url http://172.26.163.38:8084 --upload-id aeb9bc9f-3a30-439d-af9a-894e150a1673
{
    "Parts": [
        {
            "PartNumber": 1,
            "LastModified": "2022-08-24T06:55:28+00:00",
            "ETag": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
            "Size": 0
        }
    ],
    "ChecksumAlgorithm": null,
    "Initiator": {
        "ID": "NdpYiJu6LXb9yATdrjijm8sko69BDJUsLz",
        "DisplayName": "NdpYiJu6LXb9yATdrjijm8sko69BDJUsLz"
    },
    "Owner": {
        "DisplayName": "NdpYiJu6LXb9yATdrjijm8sko69BDJUsLz",
        "ID": "NdpYiJu6LXb9yATdrjijm8sko69BDJUsLz"
    },
    "StorageClass": null
}
```
**Logs:**
```
Aug 24 06:58:32 az neofs-s3-gw[5874]: 2022-08-24T06:58:32.350Z        info        api/router.go:167        call method        {"status": 200, "request_id": "89a92467-d341-4056-bc07-cec8bfa93595", "method": "UploadPart", "description": "OK"}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.923Z        INFO        sending PrepareRequest        {"height": 61166, "view": 0}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.925Z        INFO        received PrepareResponse        {"validator": 0}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.925Z        INFO        received PrepareResponse        {"validator": 1}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.925Z        INFO        sending Commit        {"height": 61166, "view": 0}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.925Z        INFO        received PrepareResponse        {"validator": 3}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.925Z        INFO        received Commit        {"validator": 3}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.926Z        INFO        received Commit        {"validator": 0}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.926Z        INFO        approving block        {"height": 61166, "hash": "c19151b1d99570719d58a2116c24b6dbf5c9595d5adef253f5456d7f2869e72b", "tx_count": 0, "merkle": "0000000000000000000000000000000000000000000000000000000000000000", "prev": "649708bace2fef633b312304b3a679dfdfb214bf675fcf01667713c533700ac1"}
Aug 24 06:58:32 az neogo-morph-cn[5233]: 2022-08-24T06:58:32.926Z        INFO        initializing dbft        {"height": 61167, "view": 0, "index": 2, "role": "Backup"}
Aug 24 06:58:33 az neofs-s3-gw[5874]: 2022-08-24T06:58:33.294Z        info        api/router.go:167        call method        {"status": 200, "request_id": "254a6efd-a261-4ca1-a397-20a655c8a613", "method": "UploadPart", "description": "OK"}
Aug 24 06:58:33 az neogo-morph-cn[5233]: 2022-08-24T06:58:33.376Z        INFO        persisted to disk        {"blocks": 1, "keys": 20, "headerHeight": 61166, "blockHeight": 61166, "took": "7.405842ms"}
Aug 24 06:58:33 az neofs-s3-gw[5874]: 2022-08-24T06:58:33.635Z        info        api/router.go:167        call method        {"status": 200, "request_id": "5b683e0c-6536-410a-9be4-65c65879ec31", "method": "UploadPart", "description": "OK"}
```
**Product versions:**
```
s3 gateway
Version: v0.23.0-36-g3b343d1-dirty
GoVersion: go1.18.4

NeoFS Storage node
Version: v0.31.0
GoVersion: go1.18.4

NeoGo
Version: 0.99.1
GoVersion: go1.18.4
```
**s3 gateway config:**
```
default_policy: REP 1 IN X CBF 1 SELECT 1 FROM * AS X
listen_address: 0.0.0.0:8084
logger:
  level: debug
max_clients_count: 600
max_clients_deadline: 60s
peers:
  '0':
    address: node1.neofs:8080
    priority: '1'
    weight: '1'
  '1':
    address: node2.neofs:8080
    priority: '2'
    weight: '0.25'
  '2':
    address: node3.neofs:8080
    priority: '2'
    weight: '0.25'
  '3':
    address: node4.neofs:8080
    priority: '2'
    weight: '0.25'
pool_error_threshold: 100
pprof:
  address: localhost:8085
  enabled: true
prometheus:
  address: localhost:8086
  enabled: true
resolve_order:
- nns
rpc_endpoint: http://node1.neofs:40332
tree:
  service: 172.26.163.38:8080
wallet:
  address: ''
  passphrase: ''
  path: /etc/neofs/s3/wallet.json
```

### PR DESCRIPTION
**Pull request by KirillovDenis**
close #672 
Signed-off-by: Denis Kirillov <denis@nspcc.ru>

### CODE REVIEW

File: api/handler/copy.go [hunk 1]
```diff
@@ -21,14 +22,18 @@ type copyObjectArgs struct {
 
 const replaceMetadataDirective = "REPLACE"
 
+var (
+	copySourceMatcher = auth.NewRegexpMatcher(regexp.MustCompile(`^/?(?P<bucket_name>[a-z0-9.\-]{3,63})/(?P<object_name>.+)$`))
+)
```
Comment (alexvanin) [inline api/handler/copy.go:27]: Single var declaration would look nicer without brackets, just some styling.

File: api/router.go [hunk 2]
```diff
@@ -219,7 +219,7 @@ func Attach(r *mux.Router, domains []string, m MaxClients, h Handler, center aut
 		bucket.Methods(http.MethodHead).Path("/{object:.+}").HandlerFunc(
 			m.Handle(metrics.APIStats("headobject", h.HeadObjectHandler))).Name("HeadObject")
 		// CopyObjectPart
-		bucket.Methods(http.MethodPut).Path("/{object:.+}").HeadersRegexp(hdrAmzCopySource, ".*?(\\/|%2F).*?").HandlerFunc(m.Handle(metrics.APIStats("uploadpartcopy", h.UploadPartCopy))).Queries("partNumber", "{partNumber:[0-9]+}", "uploadId", "{uploadId:.*}").
+		bucket.Methods(http.MethodPut).Path("/{object:.+}").HeadersRegexp(hdrAmzCopySource, ".*").HandlerFunc(m.Handle(metrics.APIStats("uploadpartcopy", h.UploadPartCopy))).Queries("partNumber", "{partNumber:[0-9]+}", "uploadId", "{uploadId:.*}").
```
Comment (alexvanin) [inline api/router.go:222]: I think you can use `.Headers(hdrAmzCopySource)` without regexp now?

**Review by alexvanin — COMMENTED** [review]
So the reason to do regexp in handler instead of a router is that we use it for splitting, right?

**Comment by KirillovDenis** [comment]
> So the reason to do regexp in handler instead of a router is that we use it for splitting, right?

No. The reason is that if regexep in `CopyObject` route fails (because of invalid `x-amz-copy-source` header) we'll go into `PutObject`. But this isn't what we want.

File: api/router.go (hunk 2 above)
Comment (KirillovDenis) [reply api/router.go:222]: Actually it will be `.Headers(hdrAmzCopySource, "")

**Review by alexvanin — APPROVED** [review]

**Pull request merged**